blockchain/fullblocktests/generate.go

```diff
@@ -1695,6 +1811,111 @@ func Generate(includeLargeReorg bool) (tests [][]TestInstance, err error) {
 	})
 	rejected(blockchain.ErrStakeTxInRegularTree)
 
+	// Create block with an invalid coinbase transaction.
+	//
+	//   ... -> brs3(14)
+	//                  \-> bmf17(15)
+	g.SetTip("brs3")
+	g.NextBlock("bmf17", outs[15], ticketOuts[15], func(b *wire.MsgBlock) {
+		b.Transactions[0].TxIn[0].BlockHeight++
+	})
+	rejected(blockchain.ErrBadCoinbaseFraudProof)
+
+	// Create block with an invalid transaction block index.
+	//
+	//   ... -> brs3(14)
+	//                  \-> bmf18(15)
+	g.SetTip("brs3")
+	g.NextBlock("bmf18", outs[15], ticketOuts[15], func(b *wire.MsgBlock) {
+		txOut := chaingen.MakeSpendableOut(b, 1, 0)
+		tx := g.CreateSpendTx(&txOut, lowFee)
+		tx.TxIn[0].BlockIndex++
+		b.AddTransaction(tx)
+	})
+	rejected(blockchain.ErrFraudBlockIndex)
+
+	// Create block with an invalid transaction block height.
+	//
+	//   ... -> brs3(14)
+	//                  \-> bmf19(15)
+	g.SetTip("brs3")
+	g.NextBlock("bmf19", outs[15], ticketOuts[15], func(b *wire.MsgBlock) {
+		txOut := chaingen.MakeSpendableOut(b, 1, 0)
+		tx := g.CreateSpendTx(&txOut, lowFee)
+		tx.TxIn[0].BlockHeight++
+		b.AddTransaction(tx)
+	})
+	rejected(blockchain.ErrFraudBlockHeight)
+
+	// Create block with an invalid transaction amount.
+	//
+	//   ... -> brs3(14)
+	//                  \-> bmf20(15)
+	g.SetTip("brs3")
+	g.NextBlock("bmf20", outs[15], ticketOuts[15], func(b *wire.MsgBlock) {
+		txOut := chaingen.MakeSpendableOut(b, 1, 0)
+		tx := g.CreateSpendTx(&txOut, lowFee)
+		tx.TxIn[0].ValueIn--
+		b.AddTransaction(tx)
+	})
+	rejected(blockchain.ErrFraudAmountIn)
+
+	// Create block with an expired transaction.
+	//
+	//   ... -> brs3(14)
+	//                  \-> bmf21(15)
+	g.SetTip("brs3")
+	g.NextBlock("bmf21", outs[15], ticketOuts[15], func(b *wire.MsgBlock) {
+		txOut := chaingen.MakeSpendableOut(b, 1, 0)
+		tx := g.CreateSpendTx(&txOut, lowFee)
+		tx.Expiry = 20
```

Comment:
This should be 1 so it is the exact expiration period of the transaction itself to help detect off by ones in the code.  There should also ideally be another test added that sets the expiry of `b.Transactions[1]` to exactly one more than the height at which it is included to ensure it is accepted.  Obviously that would cause another accepted block and thus can't go here, but such a test should be added.